File: frontend/src/app/features/caisse/services/operation-caisse.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { map, Observable } from 'rxjs';
import { API_BASE_URL } from '../../../core/services/api.config';
import { OperationCaisseRequest } from '../models/operation-caisse-request';
import { OperationCaisseResponse } from '../models/operation-caisse-response';
import { Page } from '../../../shared/models/page.model';
import { JournalCaisseResponse } from '../models/journal-caisse-response';
import { JournalCaisseFilter } from '../models/journal-caisse-filter';

@Injectable({
  providedIn: 'root'
})
export class OperationCaisseService {
  private readonly apiUrl = `${API_BASE_URL}/operations-caisse`;

  constructor(private http: HttpClient) {}

  enregistrer(request: OperationCaisseRequest): Observable<OperationCaisseResponse> {
    return this.http.post<OperationCaisseResponse>(this.apiUrl, request);
  }

  // PHASE 3B: Paginated version
  getAllPaginated(page: number = 0, size: number = 10): Observable<Page<OperationCaisseResponse>> {
    return this.http.get<Page<OperationCaisseResponse>>(`${this.apiUrl}?page=${page}&size=${size}`);
  }

  // Backward compatibility: keep original method
  getAll(): Observable<OperationCaisseResponse[]> {
    return this.http.get<Page<OperationCaisseResponse> | OperationCaisseResponse[]>(this.apiUrl).pipe(
      map((response) => Array.isArray(response) ? response : (response?.content ?? []))
    );
  }

  getBySession(sessionId: number): Observable<OperationCaisseResponse[]> {
    return this.http.get<OperationCaisseResponse[]>(`${this.apiUrl}/session/${sessionId}`);
  }

  getByCaisse(caisseId: number): Observable<OperationCaisseResponse[]> {
    return this.http.get<OperationCaisseResponse[]>(`${this.apiUrl}/caisse/${caisseId}`);
  }

  getJournalCaisse(
    filters: JournalCaisseFilter = {},
    page: number = 0,
    size: number = 20,
    sort: string = 'dateOperation,desc'
  ): Observable<Page<JournalCaisseResponse>> {
    let params = new HttpParams()
      .set('page', String(page))
      .set('size', String(size))
      .set('sort', sort);

    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params = params.set(key, String(value));
      }
    });

    return this.http.get<Page<JournalCaisseResponse>>(`${this.apiUrl}/journal`, { params });
  }
}

File: backend/src/main/java/com/mini/credit/service/OperationCaisseService.java
package com.mini.credit.service;

import com.mini.credit.dto.caisse.OperationCaisseRequest;
import com.mini.credit.dto.caisse.OperationCaisseResponse;
import com.mini.credit.dto.caisse.JournalCaisseFilterRequest;
import com.mini.credit.dto.caisse.JournalCaisseResponse;
import com.mini.credit.dto.caisse.RequalificationNatureFinancementRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface OperationCaisseService {

    OperationCaisseResponse enregistrer(OperationCaisseRequest request);

    OperationCaisseResponse enregistrerDepuisCollecteValidee(OperationCaisseRequest request);

    List<OperationCaisseResponse> getBySession(Long sessionId);

    List<OperationCaisseResponse> getByCaisse(Long caisseId);

    List<OperationCaisseResponse> getAll();  // Keep for backward compatibility
    Page<OperationCaisseResponse> getAll(Pageable pageable);  // PHASE 3B: New paginated version

    Page<JournalCaisseResponse> getJournal(JournalCaisseFilterRequest filter, Pageable pageable);

    OperationCaisseResponse requalifierNatureFinancement(Long operationId, RequalificationNatureFinancementRequest request);
}